### PR TITLE
[SPARK-11215] [ML] Add multiple columns support to StringIndexer

### PR DESCRIPTION
**Pull request by yanboliang**
Add multiple columns support to `StringIndexer`, then users can transform multiple input columns to multiple output columns simultaneously.

### CODE REVIEW

**Comment by yanboliang** [comment]
Jenkins, test this please.

**Comment by yanboliang** [comment]
Because of the multiple columns ```StringIndexer``` use ```Aggregate``` rather than ```countByValue``` to compute distinct value count, if two or more values has the same count, it will has indeterminate order.
So 1) binary classification label column may be indexed to different result(0, 1 or 1, 0); 2) ```OneHotEncoder``` will drop the last category in the encoded vector by default, if there are more than one value can be drop, it will indeterminate drop which one in this proposal.
I don't think we need to keep the order restriction produced by ```countByValue``` which may lead poor performance in the ```Aggregate``` implementation, so I disabled some test cases.
If my proposal work well, I can enable and update these test cases.

**Comment by yanboliang** [comment]
Jenkins, test this please.

**Comment by yanboliang** [comment]
Because of the multiple columns ```StringIndexer``` use ```Aggregate``` rather than ```countByValue``` to compute distinct value count, if two or more values has the same count, it will has indeterminate order.
So 
1) binary classification label column may be indexed to different result(0, 1 or 1, 0); 
2) ```OneHotEncoder``` will drop the last category in the encoded vector by default, if there are more than one value can be drop, it will indeterminate drop which one in this proposal.
I don't think we need to keep the order restriction produced by ```countByValue``` which may lead poor performance in the ```Aggregate``` implementation, so I disabled some test cases. If my proposal work well, I can enable and update these test cases.

File: R/pkg/inst/tests/test_mllib.R [hunk 1]
```diff
@@ -56,14 +56,3 @@ test_that("feature interaction vs native glm", {
   rVals <- predict(glm(Sepal.Width ~ Species:Sepal.Length, data = iris), iris)
   expect_true(all(abs(rVals - vals) < 1e-6), rVals - vals)
 })
-
-test_that("summary coefficients match with native glm", {
```
Comment (felixcheung) [inline R/pkg/inst/tests/test_mllib.R]: why is this removed?

Comment (yanboliang) [inline R/pkg/inst/tests/test_mllib.R]: It's not removed, just temporary disable. Because of this PR changed the semantics of ```StringIndexer``` a little that cause this test case produce indeterminate result. We should first discuss the semantics changing is necessary or not, and then we can update the test case to produce determinate result.

**Comment by BenFradet** [comment]
+1, I've been meaning to request this transformer for a while.

**Comment by BenFradet** [comment]
LGTM, I think a good follow up would be to do the same with `IndexToString`.

**Comment by rxin** [comment]
Thanks for the pull request. I'm going through a list of pull requests to cut them down since the sheer number is breaking some of the tooling we have. Due to lack of activity on this pull request, I'm going to push a commit to close it. Feel free to reopen it or create a new one.

**Comment by pkch** [comment]
What needs to happen to move this forward? This was a PR that would have been the first iteration of a significant improvement in handling of wide datasets.

**Comment by rxin** [comment]
I think @yanboliang just need to push this forward and get people to review it.

**Comment by MLnick** [comment]
@yanboliang will you be reviving this PR?

**Comment by yanboliang** [comment]
@MLnick I think this is an important feature to make ML pipeline handle large datasets elegantly. I will update/send a new PR soon and looking forward that you can help to review. Thanks!

**Comment by pramitchoudhary** [comment]
@yanboliang This is a very helpful initiative by you. Thanks for taking it up. Let me know, if you need any help for this PR.

**Comment by yanboliang** [comment]
@pramitchoudhary Yeah, lots of users vote for this feature. I will update this PR to match master in the next release cycle and looking forward to review/comment. Since Spark 2.1 is code freeze, we can only merge bug fix and docs change right now. Thanks.

**Comment by WeichenXu123** [comment]
@yanboliang I will take over this feature and create a new PR soon.

**Comment by minixalpha** [comment]
@WeichenXu123  Any activity for the new PR?

**Comment by WeichenXu123** [comment]
@minixalpha Sorry for delay. Too busy recently. But I will try to finish and commit my new PR once I get time. Thanks!

**Comment by minixalpha** [comment]
Thanks for you job! @WeichenXu123
